sdk/clientcore/http-netty4/src/main/java/io/clientcore/http/netty4/implementation/Netty4StreamingHttp2Adapter.java

```diff
@@ -0,0 +1,129 @@
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.DefaultFullHttpResponse;
+import io.netty.handler.codec.http.DefaultHttpContent;
+import io.netty.handler.codec.http.DefaultHttpResponse;
+import io.netty.handler.codec.http.DefaultLastHttpContent;
+import io.netty.handler.codec.http.FullHttpMessage;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpHeaders;
+import io.netty.handler.codec.http.HttpResponse;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http2.Http2CodecUtil;
+import io.netty.handler.codec.http2.Http2Connection;
+import io.netty.handler.codec.http2.Http2Error;
+import io.netty.handler.codec.http2.Http2EventAdapter;
+import io.netty.handler.codec.http2.Http2Exception;
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.handler.codec.http2.Http2Settings;
+import io.netty.handler.codec.http2.Http2Stream;
+import io.netty.handler.codec.http2.HttpConversionUtil;
+import io.netty.handler.codec.http2.InboundHttp2ToHttpAdapter;
+
+import static io.netty.handler.codec.http2.Http2Error.PROTOCOL_ERROR;
+import static io.netty.handler.codec.http2.Http2Exception.connectionError;
+import static io.netty.handler.codec.http2.HttpConversionUtil.addHttp2ToHttpHeaders;
+import static io.netty.handler.codec.http2.HttpConversionUtil.parseStatus;
+
+/**
+ * Implementation of {@link Http2EventAdapter} that converts HTTP/2 frames into HTTP/1.1 objects.
+ * <p>
+ * This is similar to {@link InboundHttp2ToHttpAdapter} but it doesn't buffer the entire response into a
+ * {@link FullHttpResponse}. Rather it streams frames as they arrive, allowing for more efficient memory usage.
+ */
+final class Netty4StreamingHttp2Adapter extends Http2EventAdapter {
+    private final Http2Connection connection;
+
+    Netty4StreamingHttp2Adapter(Http2Connection connection) {
+        this.connection = connection;
+    }
+
+    // TODO (alzimmer): This implementation is close but needs a way to control when WINDOWS_UPDARE frames are sent to
+    //  prevent race conditions between switching from the initial response data handling in Netty4ResponseHandler and
+    //  either eager or deferred content reading in the custom handlers.
+    //  For now, while huge responses don't need to be supported yet, use InboundHttp2ToHttpAdapter to buffer the
+    //  entire response into a FullHttpResponse.
+    @Override
+    public int onDataRead(ChannelHandlerContext ctx, int streamId, ByteBuf data, int padding, boolean endOfStream)
+        throws Http2Exception {
+        Http2Stream stream = connection.stream(streamId);
+        if (stream == null) {
+            throw connectionError(PROTOCOL_ERROR, "Data Frame received for unknown stream id %d", streamId);
+        }
+
+        int dataReadableBytes = data.readableBytes();
+        data = Unpooled.copiedBuffer(data);
```

Comment:
Does netty release the original ByteBuf before we finish reading? Is that why we need copy here? It would be good to add a comment.